crates/fluvio-cli/src/produce/mod.rs

```diff
@@ -112,71 +108,46 @@ impl ProduceOpt {
         Ok(())
     }
 
-    async fn produce_strs(&self, producer: &TopicProducer, strings: &[&str]) -> Result<()> {
-        if self.kv_mode() {
-            self.produce_key_values(producer, strings).await?;
-        } else {
-            let batch = strings.iter().map(|&s| (RecordKey::NULL, s));
-            producer.send_all(batch).await?;
-            if self.verbose {
-                for string in strings {
-                    println!("[null] {}", string);
-                }
-            }
-        }
-        Ok(())
-    }
-
-    fn kv_mode(&self) -> bool {
-        self.key_separator.is_some()
-    }
-
-    fn interactive_mode(&self) -> bool {
-        self.file.is_none() && atty::is(atty::Stream::Stdin)
-    }
-
-    async fn produce_key_values(&self, producer: &TopicProducer, strings: &[&str]) -> Result<()> {
+    async fn produce_line(&self, producer: &TopicProducer, line: &str) -> Result<()> {
         if let Some(separator) = &self.key_separator {
-            self.produce_key_value_via_separator(producer, strings, separator)
-                .await?;
-            return Ok(());
+            self.produce_key_value(producer, line, separator).await?;
+        } else {
+            producer.send(RecordKey::NULL, line).await?;
         }
 
-        Err(CliError::Other(
-            "Failed to send key-value record".to_string(),
-        ))
+        Ok(())
     }
 
-    async fn produce_key_value_via_separator(
+    async fn produce_key_value(
         &self,
         producer: &TopicProducer,
-        strings: &[&str],
+        line: &str,
         separator: &str,
     ) -> Result<()> {
-        debug!(?separator, "Producing Key/Value:");
-
-        let pairs: Vec<_> = strings
-            .iter()
-            .filter_map(|s| {
-                let pieces: Vec<_> = s.split(separator).collect();
-                if pieces.len() < 2 {
-                    error!("Failed to find separator '{}' in record '{}'", separator, s);
-                    return None;
-                }
+        let maybe_kv = line.split_once(separator);
+        let (key, value) = match maybe_kv {
+            Some(kv) => kv,
+            None => {
+                error!(
+                    "Failed to find separator '{}' in record, skipping: '{}'",
+                    separator, line
+                );
+                return Ok(());
+            }
+        };
 
-                let key: &str = pieces[0];
-                let value: String = (&pieces[1..]).join(&*separator);
-                if self.verbose {
-                    println!("[{}] {}", key, value);
-                }
-                Some((key, value))
-            })
-            .collect();
+        if self.verbose {
+            println!("[{}] {}", key, value);
+        }
 
-        producer.send_all(pairs).await?;
+        producer.send(key, value).await?;
         Ok(())
     }
 
+    fn interactive_mode(&self) -> bool {
+        self.file.is_none() && atty::is(atty::Stream::Stdin)
```

Comment:
for interactive use, set lingering time to low (like 10ms) otherwise, it may appear to be slow